src/ripple/protocol/STBase.h

```diff
@@ -31,7 +31,58 @@
 #include <utility>
 namespace ripple {
 
-enum class JsonOptions { none = 0, include_date = 1 };
+/// Note, should be treated as flags that can be | and &
+struct JsonOptions
+{
+    using underlying_t = unsigned int;
+    underlying_t value;
+
+    enum values : underlying_t {
+        none = 0u,
+        include_date = 1u,
+        disable_API_prior_V2 = 2u,
```

Comment:
There is a reason why I want to define this in an exclusive, rather than inclusive manner. The reason is that `prior_V2` is by definition _bounded_ i.e. its meaning does not change as we add new versions. So, if we add version 3 but keep our code say `disable_API_prior_V2` that simply implies that there are no changes between V2 and V3 (which should be the default for most of API, unless we want our users to hate us). On the other hand, `post_V1` is _unbounded_ - as we add new version, it becomes _implicitly_ included in the _meaning_ of `post_V1` and that is not what we want, unless we wanted to either rewrite all the call sites or tolerate potentially confusing name. Similar reason why I am not using `select_API_V2` which, although bounded, implies incompatibility with future versions (hence, again, would need to rewrite call sites as we add new versions).

Yes I know this goes against `if (apiVersion > 1)` programming pattern on which I seem to have settled, but such code can be easily (re)structured to make sense as we add new versions - e.g. you can nest `if (apiVersion > 2)` inside, or change to `switch` statement or use something else, because it's all imperative style (could be changed to OO or functional, but the same applies - your code is still saying "here's _how_ I want to achieve my goal"). Enums on the other hand are declarative (your code says "here's _what_ I want to achieve"), which makes them more constrained and powerful - and also why it is important to give them _bounded_ semantics.

Reply:
Yes, all good points. Agreed.